tools/flaky-test-reporter/slack_notification.go

```diff
@@ -0,0 +1,170 @@
+*/
+
+// slack_notification.go sends notifications to slack channels
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"net/http"
+	"net/url"
+
+	"github.com/knative/test-infra/shared/testgrid"
+)
+
+const (
+	knativeBotName          = "Knative bot"
+	slackChatPostMessageURL = "https://slack.com/api/chat.postMessage"
+	// default filter for testgrid link
+	testgridFilter = "exclude-non-failed-tests=20"
+)
+
+var (
+	// jobNameTestgridURLMap contains harded coded mapping of job name: Testgrid tab URL relative to base URL
+	jobNameTestgridURLMap = map[string]string{
+		"ci-knative-serving-continuous": "knative-serving#continuous",
+	}
+	// slackChannelsMap defines mapping of repo: slack channels
+	slackChannelsMap = map[string][]slackChannel{
+		"serving": []slackChannel{{"test", "CA1DTGZ2N"}},
+	}
+)
+
+type SlackClient struct {
+	userName  string
+	tokenStr  string
+	iconEmoji *string
+}
+
+// slackChannel contains channel logical name and Slack identity
+type slackChannel struct {
+	name, identity string
+}
+
+// newSlackClient reads token file and stores it for later authentication
+func newSlackClient(slackTokenPath string) (*SlackClient, error) {
+	b, err := ioutil.ReadFile(slackTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	return &SlackClient{
+		userName: knativeBotName,
+		tokenStr: strings.TrimSpace(string(b)),
+	}, nil
+}
+
+// postMessage does http post
+func (c *SlackClient) postMessage(url string, uv *url.Values) error {
+	resp, err := http.PostForm(url, *uv)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	t, _ := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("http response code is not 200 '%s'", string(t))
+	}
+	// response code could also be 200 if channel doesn't exist, parse response body to find out
+	var b struct {
+		OK bool `json:"ok"`
+	}
+	if err = json.Unmarshal(t, &b); nil != err || !b.OK {
+		return fmt.Errorf("response not ok '%s'", string(t))
+	}
+	return nil
+}
+
+// writeSlackMessage posts text to channel
+func (c *SlackClient) writeSlackMessage(text, channel string) error {
+	uv := &url.Values{}
+	uv.Add("username", c.userName)
+	uv.Add("token", c.tokenStr)
+	if nil != c.iconEmoji {
+		uv.Add("icon_emoji", *c.iconEmoji)
+	}
+	uv.Add("channel", channel)
+	uv.Add("text", text)
+
+	return c.postMessage(slackChatPostMessageURL, uv)
+}
+
+// getTestgridTabURL gets Testgrid URL for giving job
+func getTestgridTabURL(jobName string) (string, error) {
+	url, ok := jobNameTestgridURLMap[jobName]
+	if !ok {
+		return "", fmt.Errorf("cannot find Testgrid tab for job '%s'", jobName)
+	}
+	return fmt.Sprintf("%s/%s&%s", testgrid.BaseURL, url, testgridFilter), nil
+}
+
+// createSlackMessageForRepo creates slack message layout from RepoData
+func createSlackMessageForRepo(rd *RepoData) string {
+	flakyTests := getFlakyTests(rd)
+	message := fmt.Sprintf("(As of %s)\nActive flaky tests in repo '%s': %d",
```

Comment:
Rewording suggestion:

```suggestion
	message := fmt.Sprintf("As of %s, there are %d flaky tests in '%s'",
```